Retry email sending automation for unconfirmed subscribers [MAILPOET-6176]

oz opened the pull request:
## Description

When a subscriber isn’t already “Subscribed”, and some automation runs without any delay (to allow the subscriber to confirm their subscription), then any Send-Email actions will fail.

This PR fixes the issue by retrying the "Send Email Action" a couple of times if opt-in is required, and the subscriber has not yet confirmed their subscription.

## Code review notes

I discussed a bit of the approach with @JanJakes, and we found it was probably okay to re-schedule the "SendEmailAction" action a few times to wait for opt-in, and store state in the automation's run log. 💭 

> [!IMPORTANT]
> I'd like to get feedback on the current state of this PR, before spending a bit more time to improve the retry mechanism (see comments in code), or add more tests maybe. 😄 

## QA notes

Testing instructions:
1. Make sure signup confirmation is enabled in Settings.
2. Create a new automation with two steps, "Tag Added to Subscriber" as a trigger and "Send email" as an action. Activate it.
3. Add a new unconfirmed subscriber.
4. Add a tag to this subscriber on the subscriber edit page.
5. Wait until cron execution and observe that the automation run is in progress. Wait for some time to see that it is not advancing further.
6. Confirm the subscriber.
7. Check that the automation run is now completed.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6176](https://mailpoet.atlassian.net/browse/MAILPOET-6176)

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6176]: https://mailpoet.atlassian.net/browse/MAILPOET-6176?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:MAILPOET-6176_automations-subscribed-only)

_The latest successful build from `MAILPOET-6176_automations-subscribed-only` will be used. If none is available, the link won't work._

oz committed:
Allow SendEmailAction to retry for optin reasons

If an automation that requires opt-in is run before a subscriber can
confirm their email, re-schedule the action a couple of times before
giving up.

[MAILPOET-6176]

websupporter commented on mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendEmailAction.php:
```diff
@@ -171,40 +186,115 @@ public function run(StepRunArgs $args, StepRunController $controller): void {
 
+  private function isOptInRequired(NewsletterEntity $newsletter, SubscriberEntity $subscriber): bool {
+    $subscriberStatus = $subscriber->getStatus();
+    return ($newsletter->getType() !== NewsletterEntity::TYPE_AUTOMATION_TRANSACTIONAL && $subscriberStatus !== SubscriberEntity::STATUS_SUBSCRIBED);
```
Maybe just taste, but I find this hard to read. So much combined negativity.

i think that would make it easier for me:
```
if transactional return false;
return !subscribed
```

oz replied:
No problem with less negativity. 😄

FTR I merely extracted existing code here. 😇

websupporter commented:
Overall, I think that looks good. We talked yesterday about the retry count approach and it makes sense to me to store some kind of "retries for optin"-number.

oz committed:
Batch updates of AutomationRunLog data

oz committed:
Simplify opt-in requirement conditional

The original logic is fine. This is more for readability, with multiple
statements instead of a single condition.

oz committed:
Offset opt-in reruns from sending reruns

In automations, the send-email action can be rerun to wait for
subscriber opt-in, but also to wait for the the email scheduler (retry
sending). We use a simple counter to know how many times the action was
retried.

Both "reruns" have their own retry schedule however. Once opt-in has
been acquired, we must not mix the opt-in retry count, with the sending
retry count.

alex-mpoet commented on mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendEmailAction.php:
```diff
@@ -171,40 +186,115 @@ public function run(StepRunArgs $args, StepRunController $controller): void {
-      // run #N: check/sync sending status with the automation step
+      // Re-running for opt-in?
+      $state = $this->getRunLogData($args);
+      if (array_key_exists(self::WAIT_OPTIN, $state) && $state[self::WAIT_OPTIN] === 1) {
```
Not a strict requirement, but if `WAIT_OPTIN` is treated as a boolean, this could be simplified as `if (!empty($state[self::WAIT_OPTIN])) {`.

alex-mpoet reviewed (commented):
@oz I have an issue when trying to test this. After the `{"wait_optin":1}` flag gets saved to the DB, it is immediately wiped by `StepRunLogger->logProgress()` and other `StepRunLogger` methods, so it is absent during the next automation run. As a result, the code [here](https://github.com/mailpoet/mailpoet/pull/5967/files#diff-1e596733ccdea88d2ded67e24bca2d74c50867d0fe30c9137a81063fa64d7e3cR206) thinks the subscriber is confirmed and email scheduling fails like it did before this fix. Does it work for you?

oz commented:
> @oz I have an issue when trying to test this. After the `{"wait_optin":1}` flag gets saved to the DB, it is immediately wiped by `StepRunLogger->logProgress()` and other `StepRunLogger` methods, so it is absent during the next automation run. As a result, the code [here](https://github.com/mailpoet/mailpoet/pull/5967/files#diff-1e596733ccdea88d2ded67e24bca2d74c50867d0fe30c9137a81063fa64d7e3cR206) thinks the subscriber is confirmed and email scheduling fails like it did before this fix. Does it work for you?

Hi @alex-mpoet! Can you detail the steps to reproduce this?

alex-mpoet commented:
@oz

> Hi @alex-mpoet! Can you detail the steps to reproduce this?

I just tested the basic scenario described in the PR:
1. Make sure signup confirmation is enabled in Settings.
2. Create a new automation with two steps, "Tag Added to Subscriber" as a trigger and "Send email" as an action. Activate it.
3. Add a new unconfirmed subscriber.
4. Add a tag to this subscriber on the subscriber edit page.
5. Wait until cron execution and observe the failed automation run with the "'Email failed to schedule" error.

oz commented:
> @oz
> 
> > Hi @alex-mpoet! Can you detail the steps to reproduce this?
> 
> I just tested the basic scenario described in the PR:
> 
> 1. Make sure signup confirmation is enabled in Settings.
> 2. Create a new automation with two steps, "Tag Added to Subscriber" as a trigger and "Send email" as an action. Activate it.
> 3. Add a new unconfirmed subscriber.
> 4. Add a tag to this subscriber on the subscriber edit page.
> 5. Wait until cron execution and observe the failed automation run with the "'Email failed to schedule" error.

Ah thanks @alex-mpoet! 🙏🏻 

I found the issue. Indeed, it looks like the step-runner has its own instance of a logger, and will overwrite the one created in the action. I have a fix that involves passing the a step-run logger through (e.g. using a step-run controller), but I'm not very satisfied with it yet. I'll resubmit for review later.

oz committed:
Reuse StepRunLogger to avoid overwrites

The SendEmailAction step needs to log arbitrary data to allow
retry-runs, through a StepRunLogger. We previously instanciated a logger
from the action, but this data will be overwritten by the handler right
after the action is run. This is because the handler holds a reference
to a logger already, and will overwrite any data the action tries to
write.

To work around this issue, the step handler now passes its logger
instance to the step's controller, allowing safe access from
SendEmailAction.

[MAILPOET-6176]

oz commented:
@alex-mpoet can you take another look at this? To work-around the potential overwrites, I found that I had to share the instance of the "logger" from the class that handles step-runs. I'm not sure that this is the best approach overall, but it means that the data is no longer overwritten, and the action can log its own data too.

alex-mpoet reviewed (approved):
Thanks for fixing, the retrying mechanism works now and looks good!

Aschepikov reviewed (approved):
GJ!
